### ISSUE TITLE
Error "MultiplyDescriptor: an illegal memory access was encountered" in feature matcher running on cluster

### ISSUE THREAD

**Issue by lixx2938** [colmap/colmap#68]
Hello,

I am trying to run COLMAP in my lab's cluster without display but there is still one problem for feature matcher:
the commands I used:
(1) ./src/exe/feature_extractor    --General.database_path $PROJECT_PATH/database.db    --General.image_path $PROJECT_PATH --use_gpu false
(2) ./src/exe/exhaustive_matcher    --General.database_path $PROJECT_PATH/database.db --gpu_index 0

Then when I run any matcher, I will get some repeated error output like:
Matching block [1/8, 5/8]MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
..................

my cluster's cuda version is 7.5 and output of graphics card is GTX Titan X:
| NVIDIA-SMI 361.42     Driver Version: 361.42         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 0000:02:00.0     Off |                  N/A |
| 22%   42C    P0    73W / 250W |     23MiB / 12287MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

However, I have no problem running CUDA version of sparse SfM and dense MVS so I don't think it's due to CUDA. Does anyone know if I need to change or add some flags in cmake or something wrong with my commands? I have been stuck with this problem for long time and I would appreciate if someone could help me figure it out.

Thanks

**Comment by lixx2938** [comment]
I also found this error didn't occurs at the very beginning of matching. For example, it only occurs after 5 image matching:
Matching image [1/40] in 3.015s
Matching image [2/40] in 0.705s
Matching image [3/40] in 0.717s
Matching image [4/40] in 0.702s
Matching image [5/40] in 0.459s
Matching image [6/40]MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered

Has anyone successfully run CUDA (not OpenGL version) feature matching without error?

**Comment by ahojnnes** [comment]
Yes, I am running it with the CUDA version on many machines and did not encounter the problem. Can you send me the database?

**Comment by lixx2938** [comment]
Hi,
You can download my test 40 images and database after I run matching (with "MultiplyDescriptor: an illegal memory access was encountered" error) at
https://github.com/lixx2938/CV-1

You can try to run colmap on your machine and compare the database difference. Since I have no problem with sparse and dense part, I don't think this is because of CUDA or GPU; Also, I use GTX Titan X GPU and CUDA 7.5 and I believe you have similar settings as mine.

Thanks

**Comment by lixx2938** [comment]
FYI, my commands to run are:
./src/exe/feature_extractor    --General.database_path $PROJECT_PATH/database.db    --General.image_path $PROJECT_PATH/images --use_gpu false

./src/exe/vocab_tree_matcher --General.database_path $PROJECT_PATH/database.db --VocabTreeMatchOptions.vocab_tree_path  /phoenix/S2/zl548/vocab_tree/vocab_tree-1048576.bin --MatchOptions.gpu_index 0

**Comment by lixx2938** [comment]
Hi,

I just checked feature matcher CUDA code and I found in code "ProgramCU.cu", in the method "ProgramCU::MultiplyDescriptor", the first two lines:
"	int num1 = des1->GetImgWidth() / 8;
	int num2 = des2->GetImgWidth() / 8;"

I found both num1 and num2 are zero. I think this is the reason that it cause illegal memory access because grid and block size of CUDA kernel is almost zero is these two numbers are 0.

Do you have any idea which part might be wrong in the code (or where I should check in your code), I think my program somehow didn't read image size into SIFTGPU class written by Changchang WU.

Thanks

**Comment by lixx2938** [comment]
Hi,

Could you tell me your GPU driver version when you tested on your machine. I searched the problem on the website and someone said it might be due to Nvidia driver incompatibility.

Thanks,

**Comment by ahojnnes** [comment]
Ubuntu 14.04.2 LTS

```
+------------------------------------------------------+                       
| NVIDIA-SMI 355.11     Driver Version: 355.11         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 0000:02:00.0     Off |                  N/A |
| 22%   58C    P0    72W / 250W |     23MiB / 12287MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TIT...  Off  | 0000:03:00.0     Off |                  N/A |
| 22%   59C    P0    71W / 250W |     23MiB / 12287MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX TIT...  Off  | 0000:82:00.0     Off |                  N/A |
| 22%   54C    P0    73W / 250W |     23MiB / 12287MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce GTX TIT...  Off  | 0000:83:00.0     Off |                  N/A |
|  0%   51C    P0    55W / 250W |     23MiB / 12287MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**Comment by lixx2938** [comment]
Thank you!

By the way, do you know approximately how long it will take for you to run CPU feature matching for each image if we use vocabulary tree matching with 100 nearest neighborhoods since right now it seems that CPU feature matching on my cluster is extremely slow (500s per images)?

Thank you

**Comment by lixx2938** [comment]
Hi,

I just checked the CUDA code of matching, it seems that when we allocate memory of dot product matrix which is the product of number of features in two images, and bind GPU memory to texture memory, it will exceeded maximum limit of texture memory. I used default 32768 for maximum matching and it will crashed. Did you use default setting of maximum number of matching? If so, it should be impossible to run CUDA matching because it should exceed limit of texture memory.

Thanks,
